pf/proto/block.go

```diff
@@ -0,0 +1,134 @@
+func (b blockResource) DeprecationMessage() string { return deprecated(b.block.Deprecated) }
+
+type blockMap struct{ block *tfprotov6.SchemaBlock }
+
+func (m blockMap) Len() int { return len(m.block.Attributes) + len(m.block.BlockTypes) }
+
+func (m blockMap) Get(key string) shim.Schema { return getSchemaMap(m, key) }
+
+func (m blockMap) GetOk(key string) (shim.Schema, bool) {
+	for _, a := range m.block.Attributes {
+		if a.Name == key {
+			return attribute{*a}, true
+		}
+	}
+
+	for _, b := range m.block.BlockTypes {
+		if b.TypeName == key {
+			return blockSchema{*b}, true
+		}
+	}
+
+	return nil, false
+}
+
+func (m blockMap) Range(each func(key string, value shim.Schema) bool) {
+	for _, a := range m.block.Attributes {
+		if !each(a.Name, attribute{*a}) {
+			return
+		}
+	}
+
+	for _, b := range m.block.BlockTypes {
+		if !each(b.TypeName, blockSchema{*b}) {
+			return
+		}
+	}
+}
+
+func (m blockMap) Set(key string, value shim.Schema) {
+	switch v := value.(type) {
+	case attribute:
+		v.attr.Name = key
+		m.block.Attributes = append([]*tfprotov6.SchemaAttribute{&v.attr}, m.block.Attributes...)
+	case blockSchema:
+		v.block.TypeName = key
+		m.block.BlockTypes = append([]*tfprotov6.SchemaNestedBlock{&v.block}, m.block.BlockTypes...)
+	default:
+		contract.Failf("Must set an %T, found %T", v, value)
+	}
+}
+
+func (m blockMap) Delete(key string) {
+	// Because blocks are attributes are disjoint, we can just attempt to delete from
+	// both.
+	m.block.Attributes = filter(m.block.Attributes, func(a *tfprotov6.SchemaAttribute) bool {
+		return a.Name != key
+	})
+	m.block.BlockTypes = filter(m.block.BlockTypes, func(b *tfprotov6.SchemaNestedBlock) bool {
+		return b.TypeName != key
+	})
+}
+
+func (m blockMap) Validate() error { return nil }
+
+type blockSchema struct{ block tfprotov6.SchemaNestedBlock }
+
+func (m blockSchema) Type() shim.ValueType                { return shim.TypeMap }
+func (m blockSchema) Optional() bool                      { return false }
+func (m blockSchema) Required() bool                      { return false }
+func (m blockSchema) Default() interface{}                { return nil }
+func (m blockSchema) DefaultFunc() shim.SchemaDefaultFunc { return nil }
+func (m blockSchema) DefaultValue() (interface{}, error)  { return nil, nil }
+func (m blockSchema) Description() string                 { return m.block.Block.Description }
+func (m blockSchema) Computed() bool                      { return false }
+func (m blockSchema) ForceNew() bool                      { return false }
+func (m blockSchema) StateFunc() shim.SchemaStateFunc     { return nil }
+func (m blockSchema) Elem() interface{}                   { return blockResource{block: m.block.Block} }
+func (m blockSchema) MaxItems() int                       { return int(m.block.MaxItems) }
+func (m blockSchema) MinItems() int                       { return int(m.block.MinItems) }
+func (m blockSchema) ConflictsWith() []string             { return nil }
```

Comment:
I guess these are the bits which are inaccessible through GRPC? Are default inaccessible here too?

Reply:
Yeah no luck with detecting these over gRPC. We can just detect validation failures.

Reply:
Yep.